Big PR merging entire project into master

gwaygenomics opened the pull request:
Below is a concise summary of the project's structure, which will aid in review.

I am naming @huqiwen0313 as the main reviewer, since you has expertise in sequencing pipelines. @danich1, your review on the python scripts and jupyter notebooks, and overall documentation, will also be very valuable. Thanks!

## Summary of Files Requiring Review

|Suggested Order | File | Purpose |
| :------------- | :--- | :------ |
| 1 | [README.md](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-04c6e90faac2675aa89e2176d2eec7d8) | provides the overview of the project |
| 2 | [LICENSE.md](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-37854d19817c792316d481f5beb93cc7) | provide reuse guidelines and establish copyright |
| 3 | [setup_environment.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-fbc714928d20dd4613c5078bc87e67e3) | script that initializes computational environment  |
| 4 | [install.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-3fbb47e318cd8802bd325e7da9aaabe8) | script to prep pipeline for use |
| 5 | [wes_pipeline.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-173531f56e4892bbcfe28f63237eb782) | command in order of which commands to run |
| 6 | [util/command_wrapper.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-07762d567477a18e05ed43c861573e25) | probably most important script, actually specifies the pipeline commands |
| 7 | [scripts/1.run_fastqc.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-39a78ba4827234feec18c25b8d3d8b86) | facilitate fastqc command for each sample |
| 8 | [scripts/2.run_trimgalore.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-de7f4f1808b894973e2160c201a57dce) | facilitate trimgalore command for each sample |
| 9 | [scripts/3.run_bwa.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-5725c37c8b7fab9449ae73a901202e0c) | facilitate bwa mem command for each sample |
| 10 | [scripts/4.run_samtools.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-bec3f88441c8d186fb7366bec84a8f5e) | facilitate samtools command for each sample |
| 11 | [scripts/5.disambiguate_species.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-b7bea9540ae9ea69dd1636a50d83e7f3) | disambiguation of species reads |
| 12 | [scripts/6.variant_calling.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a1a4f935212858594e4598a30c067b7e) | calls variants |
| 13 | [scripts/7.annotate_variants.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-d64a0872971ab69dbf27734d1d233d2e) | attaches annotation to variants |
| 14 | [disambiguate_reads.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a482e0503b4c262b9f3341364e6eb824) | notebook to survey and visualize species separation step |
| 15 | [filter_variants.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-9b7f6f40d1be458903e3881b48a963cd) | notebook to visualize variants across replicates |
| 16 | [oncoprint_data.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-76d61d98bb124b510d015c11528f9f51) | data processing to get files in format ready for oncoprint viz |
| 17 | [scripts/viz/oncoprint_similarity.R](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-bcf3d63046daa38565d7ce54dd5057f9) | outputs oncoprint diagrams and similarity viz |
| 18 | [scripts/viz/variant_overlaps.R](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-263cca28ca70731c6f04840bcbfe65be) | outputs venn diagrams of variant calling |
| 19 | [util/schedule.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-fe30fe6d15181af712b152399fd398d5) | helper function to submit job to Dartmouth compute cluster |
| 20 | [discovery_variables.yml](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-99aae3d88e8f4d7d1e209c53279c3ade) | file that specifies global variables for pipeline tools |
| 21 | [environment.yml](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-b7cb75ca59993a879e29bed8cd8469b8) | package versions for computational environment |
| 22 | [.gitignore](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a084b794bc0759e7a6b77810e01874f2) | files and folders to preclude from commiting |

There are also several results and figures included below. They should be reviewed as well, but I did not add them above.

gwaygenomics commented:
I am also tagging @danich1 for supplementary python review. @danich1 - I can explain the project to you in more detail and how your review would supplement @huqiwen0313 's. Maybe tomorrow after closing scrum?

As I mentioned above, many of the files are autogenerated and do not need review. I will write a concise summary table (including links) of which files need review and what they're doing.

gwaygenomics commented:
@huqiwen0313 and @danich1 - I added a summary of the important files above. Please let me know if you have any questions, thanks!

danich1 commented on scripts/1.run_fastqc.py:
```diff
@@ -0,0 +1,55 @@
+config = args.config_yaml
+walltime = args.walltime
+nodes = str(args.nodes)
+cores = str*args.cores)
```
* -> (

gwaygenomics replied:
ughh....sorry

gwaygenomics commented:
Thanks @danich1 

> One question I have is why wes_pipeline.sh fully commented out?

Because its a pipeline that needs to be run piece-wise. The script submits jobs and currently, job status needs to be monitored manually. In many pipeline examples I've seen (not too many actually since most do not release code publicly!) this has been the case. Definitely open to discussing alternative strategies, however!

danich1 commented:
Before I get into more code revisions, please refer to issue #25 on my idea to change the design of this pipeline a little bit.

huqiwen0313 commented on wes_pipeline.sh:
```diff
@@ -0,0 +1,129 @@
+# that are no longer best-practices for GATK HaplotypeCaller pipelines
+# (see https://software.broadinstitute.org/gatk/blog?id=7847)
+
+# Assign read groups
```
what are read groups ?

gwaygenomics replied:
I think [this discussion](https://gatkforums.broadinstitute.org/gatk/discussion/6472/read-groups) best describes read groups.

They basically ID reads output from different sequencing runs. For my purposes, there is only one read group per BAM, so the command is only necessary because MuTect2 requires them

huqiwen0313 commented on wes_pipeline.sh:
```diff
@@ -0,0 +1,129 @@
+# python scripts/4.run_samtools.py --command 'index_bam' \
+#        --data_dir 'processed/bam_rmdup' --output_dir 'processed/bam_rmdup' \
+#        --walltime '03:30:00' --nodes 1 --cores 4 
+
```
A summary script that summarize all the reads statistics maybe helpful.

gwaygenomics replied:
That would be great - can you point me to a tool or script that would help me to do this?

huqiwen0313 replied:
If you have bam files, you can use bamtool stats for the summary of reads statistics. RSeQC is another tool can be used (http://rseqc.sourceforge.net/#read-distribution-py).

gwaygenomics replied:
Thanks @huqiwen0313 ! I will look into this. I agree it will be good to do. I add issue #27 to add to the repository in the future

huqiwen0313 commented on wes_pipeline.sh:
```diff
@@ -0,0 +1,129 @@
+#       --data_dir 'processed/bam_sort_position' \
+#       --output_dir 'processed/bam_rmdup' \
+#       --walltime '03:30:00' --nodes 1 --cores 4
+
```
Not sure if this will generate big difference.  picar tool is another tool to remove duplicates. It can also specify which types of duplicates you want to move.

gwaygenomics replied:
yeah, I did notice `picard` too. Do you have a specific tool preference?

I am not too concerned about this step - it appears that ["PRC duplicate removal has minimal effect on the accuracy of subsequent variant calls"](https://doi.org/10.1186/s12859-016-1097-3)

huqiwen0313 replied:
I don't have preference, the algorithm of picar tool is a slightly different from samtools. I don't think it will make big difference either.

gwaygenomics commented on README.md:
```diff
@@ -1 +1,79 @@
+**Figure 1A** describes the technical replicates and data-types available across
+tumor and mouse passages. **Figure 1B** outlines our analysis pipeline from quality
+control processing raw reads, to alignment and removal of mouse reads, to
+calling and annotating variants.
```
> For calling variants, you remove reads that aligned to both human and mouse, right ?

That is correct!

> How many percentage of reads that aligned to both human and mouse ?

Quite a small fraction align to both (i.e. called "ambiguous"). I visualize the distribution of mouse/human/ambiguous reads in [this figure](https://github.com/greenelab/pdx_exomeseq/blob/develop/figures/disambiguate_results.pdf).  The data that form this figure is [here](https://github.com/greenelab/pdx_exomeseq/blob/develop/results/full_disambiguate_summary.tsv).

> Some of those reads may align to homologous regions (conserved regions) between human and mice genome. Those maybe also interesting regions to look at.

If they are both of high quality, the reads will almost certainly align to highly conserved regions! Do you know of any resource that describes variation in homologous regions? Specifically for cancer-associated variants?

huqiwen0313 replied:
The reason I think these regions would be interesting is if variants appear in the conserved regions, they are more likely to be functionally related. I did not see a database list all of the cancer-associated variance in the conserved regions, but there are some related papers (e.g. https://www.nature.com/articles/srep22124, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3919555/).

gwaygenomics replied:
Thanks @huqiwen0313 - I agree this could be important to look at. I added this idea as issue #26. It is beyond the scope of this PR, but will be done in the future

gwaygenomics commented on oncoprint_data.ipynb:
```diff
@@ -0,0 +1,914 @@
+{
+ "cells": [
```
yeah, that would be cool to visualize too. This seems related to #27 and can be done in the next PR. Thanks!

gwaygenomics commented on oncoprint_data.ipynb:
```diff
@@ -0,0 +1,914 @@
+       "5584                  .     synonymous SNV   \n",
+       "\n",
+       "                                       AAChange.refGene   ...    \\\n",
+       "X1                                                        ...     \n",
```
I agree - it will tell us 2 different measurements of mutation rate. What comparisons do you have in mind? Some sort of visualization?

huqiwen0313 replied:
I am thinking of visualize the distribution of different type of mutations in the samples and in the genes you are interested. Something similar to this figure: http://www.pnas.org/content/109/10/3879/F1.expansion.html. Not sure if it is feasible in your project.

gwaygenomics replied:
I currently have oncoprint diagrams, but only with mutations listed. This could change if we call copy number variants. To be determined

gwaygenomics commented on filter_variants.ipynb:
```diff
@@ -0,0 +1,546 @@
+   ],
+   "source": [
+    "p = (\n",
+    "    gg.ggplot(filter_melt_df,\n",
```
👍 sounds good, I will add the percentage above each bar in the next commit. I had tried doing this before, but it was somewhat tricky.

gwaygenomics replied:
Just realized I didn't do this yet. I added #28 as something to potentially add later on, after speaking with collaborators.

gwaygenomics commented:
|Suggested Order | File | Purpose |
| :------------- | :--- | :------ |
| 1 | [README.md](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-04c6e90faac2675aa89e2176d2eec7d8) | provides the overview of the project |
| 2 | [LICENSE.md](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-37854d19817c792316d481f5beb93cc7) | provide reuse guidelines and establish copyright |
| 3 | [setup_environment.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-fbc714928d20dd4613c5078bc87e67e3) | script that initializes computational environment  |
| 4 | [install.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-3fbb47e318cd8802bd325e7da9aaabe8) | script to prep pipeline for use |
| 5 | [wes_pipeline.sh](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-173531f56e4892bbcfe28f63237eb782) | command in order of which commands to run |
| 6 | [pdx_exomeseq.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-228b50780f2358820770ebdf41df4fe2) | **UPDATED** processes all commands for all samples and submits jobs to cluster  |
| 7 | [scripts/annotate_variants.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a5ed32f5e8c3dfe612f5d27b66d80f25) | attaches annotation to variants |
| 8 | [disambiguate_reads.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a482e0503b4c262b9f3341364e6eb824) | notebook to survey and visualize species separation step |
| 9 | [filter_variants.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-9b7f6f40d1be458903e3881b48a963cd) | notebook to visualize variants across replicates |
| 10 | [oncoprint_data.ipynb](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-76d61d98bb124b510d015c11528f9f51) | data processing to get files in format ready for oncoprint viz |
| 11 | [scripts/oncoprint_similarity.R](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-7d6048a758a274dab7c4214ebd218ed8) | outputs oncoprint diagrams and similarity viz |
| 12 | [scripts/variant_overlaps.R](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-762222df1f9c09e99e9c32e5c2f08ea9) | outputs venn diagrams of variant calling |
| 13 | [util/schedule.py](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-fe30fe6d15181af712b152399fd398d5) | helper function to submit job to Dartmouth compute cluster |
| 14 | [discovery_variables.yml](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-99aae3d88e8f4d7d1e209c53279c3ade) | file that specifies global variables for pipeline tools |
| 15 | [environment.yml](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-b7cb75ca59993a879e29bed8cd8469b8) | package versions for computational environment |
| 16 | [.gitignore](https://github.com/greenelab/pdx_exomeseq/pull/24/files#diff-a084b794bc0759e7a6b77810e01874f2) | files and folders to preclude from commiting |

gwaygenomics commented:
@huqiwen0313 and @danich1 thank you both for insightful pull request comments. :point_up: I updated the outline of the files above and included major updates and file additions. I also removed nearly all numbered scripts in favor of a single modular script as discussed in #25 

@huqiwen0313 I will perform all of your suggested pipeline enhancements (#26 and #27) in future pull requests. They are beyond the scope of the current pull request, but will be important analyses moving forward.

Ready for your review again!

danich1 commented on util/arguments.py:
```diff
@@ -0,0 +1,150 @@
+    wes_files = []
+    for path, subdirs, files in os.walk(args.input_directory):
+        for name in files:
+            if 'fasta.gz' in name or 'fq.gz' in name or 'fastq.gz' in name:
```
Could get away with `re.match(r'(fasta|fq|fastq)\.gz$', name)`, which will return None if there isn't a match at all.

gwaygenomics replied:
:+1:

danich1 commented on util/arguments.py:
```diff
@@ -0,0 +1,150 @@
+    sam_files = []
+    for path, subdirs, files in os.walk(args.input_directory):
+        for name in files:
+            if (('fq.gz.sam' in name) or ('.sam_sorted.bam' in name) or
```
Def think you should switch to regex again. I.E:
```
regex_match = r'(fq\.gz\.sam|\.sam_sorted\.bam|\.bam_sorted_fixmate\.bam|'disambiguatedSpeciesA\.bam)$'
re.match(regex_match, name)
```

gwaygenomics replied:
Hmm, I think adding it like this is starting to get a bit hard to read. It may also be harder to update if I need to make changes to this function as a result of #26 - therefore, I think I will make a list called `check_suffix` that stores all files we want and make a call `if (name in check_suffix) and ('.bai' not in name):`

danich1 commented on util/schedule.py:
```diff
@@ -0,0 +1,87 @@
+            return None
+        self.command = command
+
+    def set_command(self, command):
```
Does this command ever get called in your new design? Seems not needed anymore if you are just passing command in the initialization step

gwaygenomics replied:
True!

danich1 commented on util/schedule.py:
```diff
@@ -0,0 +1,87 @@
+    def set_command(self, command):
+        self.command = command
+
+    def set_name_command(self, name, command):
```
Is this needed anymore? Similar as above comment about `set_command()`.

danich1 commented on scripts/nbconverted/oncoprint_data.py:
```diff
@@ -0,0 +1,158 @@
+
+# Generate and save oncoprint data for consensus samples
+oncoprint_consensus_file = os.path.join('results', 'oncoprint_consensus.tsv')
+oncoprint_consensus_df = pd.DataFrame(variant_assign_consensus, index=case_id_consensus,
```
Would break this apart into:
```
oncoprint_consensus_df = (
pd.DataFrame(
  variant_assigned_consensus,
  index=case_id_consensus,
  columns=paad_genes
  )
```
Makes it easier to read.
)

gwaygenomics replied:
:+1:

danich1 commented on scripts/nbconverted/filter_variants.py:
```diff
@@ -0,0 +1,251 @@
+                       self.cosmic_variant_counts]
+        return filter_list
+    
+    def curate_summary_info(self):
```
would remove this function or change the return statement to call `summary_statistics()`

gwaygenomics replied:
nice. Yes, I will do that in the next commit!

danich1 commented on scripts/nbconverted/filter_variants.py:
```diff
@@ -0,0 +1,251 @@
+        g = sns.distplot(self.variant_df['depth'].astype(float))
+        g.set_title('Read Depth Distribution for {}'.format(self.sample_name))
+    
+    def summary_statistics(self):
```
Make this function `self.functional_counts, self.mutational_class_counts, self.depth_summary`

gwaygenomics replied:
hmmm, I'm not sure I know what you mean

gwaygenomics replied:
make separate functions for each three?

danich1 replied:
GAH! github deleting words.. anyway I was trying to say make `summary_statistics` return the three variables I mentioned above and then change `curate_summary_info` function to return a call to summary statistics

gwaygenomics replied:
ok sounds good. I ended up deleting `curate_summary_info` and renaming `summary_statistics` to `get_summary_statistics` and directly returning objects of interest instead

danich1 commented on scripts/nbconverted/filter_variants.py:
```diff
@@ -0,0 +1,251 @@
+        self.depth_summary.columns = [self.sample_name]
+
+    def curate_filter_info(self):
+        filter_list = [self.sample_name, self.all_variant_count, self.filter_min_depth_count,
```
Lets add a bit more white space and new lines to look like:
```
filter _list = [
  self.sample_name, self.all_variant_count,
  self.filter_min_depth_count, self.filter_common_var_count,
  ...
]
```

gwaygenomics replied:
:+1:

huqiwen0313 commented on scripts/oncoprint_similarity.R:
```diff
@@ -0,0 +1,126 @@
+col = c("MUT" = "#008000", "AMP" = "red", "HOMDEL" = "blue")
+
+process_oncoprint <- function(onco_file) {
+  mat <- read.table(onco_file, header = TRUE, stringsAsFactors = FALSE,
```
Can you add some description for each function ?

gwaygenomics replied:
Ah, yes. I should do that

gwaygenomics commented on oncoprint_data.ipynb:
```diff
@@ -327,8 +327,8 @@
      "text": [
-      "CPU times: user 15 s, sys: 84 ms, total: 15.1 s\n",
-      "Wall time: 15.2 s\n"
+      "CPU times: user 12.5 s, sys: 68 ms, total: 12.6 s\n",
```
Ha! This is the difference between my computer and Casey's (which I was using for about a month)

gwaygenomics commented:
Thanks @huqiwen0313 and @danich1 again for your comments. I believe I addressed them in the most recent set of commits. Ready for review again
